cmd/vm-informant/main.go

```diff
@@ -103,48 +110,106 @@ func main() {
 // We execute ourselves as a subprocess so that it's possible to appropriately cleanup after
 // termination by various signals (or an unhandled panic!). This is worthwhile because we *really*
 // don't want to leave the cgroup frozen while waiting to restart.
-func runRestartOnFailure(args []string, cleanupHooks []func()) {
+func runRestartOnFailure(ctx context.Context, args []string, cleanupHooks []func()) {
 	selfPath := os.Args[0]
-
-	minWaitDuration := time.Second * 5
+	timer := time.NewTimer(0)
+	defer timer.Stop()
 
 	for {
 		startTime := time.Now()
+		processedSignal := make(chan struct{})
+		var exitMode string
 
-		cmd := exec.CommandContext(context.TODO(), selfPath, args...)
-		cmd.Stdout = os.Stdout
-		cmd.Stderr = os.Stderr
-		err := cmd.Run()
+		func() {
+			defer close(processedSignal)
+			pctx, pcancel := context.WithCancel(context.Background())
+			defer pcancel()
+
+			cmd := exec.Command(selfPath, args...)
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+
+			err := cmd.Start()
+
+			if err == nil {
+				go func() {
+					select {
+					case <-pctx.Done():
+						return
+					case <-ctx.Done():
+						if cmd.Process == nil || cmd.ProcessState == nil {
+							return
+						}
+						if cmd.ProcessState.ExitCode() >= 0 {
+							return
+						}
+						var pid int
+						if cmd.Process != nil {
+							pid = cmd.Process.Pid
+						}
+						if err := cmd.Process.Signal(syscall.SIGTERM); err != nil {
+							klog.Infof("could not signal informant process %d: %v", pid, err)
+						}
+					}
+				}()
+
+				// this is blocking, but we should
+				// have killed the process in the
+				// wait goroutine, or the process would
+				// return normally.
+				err = cmd.Wait()
+				// stop the goroutine above, as the
+				// process has already returned.
+				pcancel()
+			}
 
-		var exitMode string
+			if err != nil {
+				// lint note: the linter's worried about wrapped errors being incorrect with switch, but
+				// this is cleaner than the alternative (via errors.As) and it's still correct because
+				// exec.Command.Wait() explicitly mentions ExitError.
+				switch err.(type) { //nolint:errorlint // see above.
+				case *exec.ExitError:
+					exitMode = "failed"
+					klog.Errorf("vm-informant exited with: %v", err)
+				default:
+					exitMode = "failed to start"
```

Comment:
I believe that if there are problems writing to the output we could have an error from wait that isn't "failed to start", having said that, I'd be inclined to drop the `exitMode` entirely and hope that the error doesn't need any more annotation.

Reply:
Fair enough, go for it